File: src/decrypt/cpp/decrypt.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <cctype>
#include <stdexcept>
#include <ranges>

constexpr std::string_view ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*_-?";
constexpr std::string_view NOISE_SET = "abcdefghijklmnopqrstuvwxyz";
constexpr char TERM = '~';
constexpr int BASE = 47;
constexpr size_t BLOCK_BYTES = 5;
constexpr size_t BLOCK_SYMBOLS = 8;
constexpr size_t CHECKSUM_PERIOD = 4;

auto build_rev(std::string_view alpha) {
    std::unordered_map<char, int> rev;
    for (size_t i = 0; i < alpha.size(); ++i) {
        rev[alpha[i]] = static_cast<int>(i);
    }
    return rev;
}

int rotation_for_block(int block_index) {
    return (block_index * 13 + 11) % BASE;
}

std::string rotate_alpha(std::string_view alpha, int rot) {
    int n = static_cast<int>(alpha.size());
    rot = ((rot % n) + n) % n;
    std::string rotated;
    rotated.reserve(alpha.size());
    for (int i = 0; i < n; ++i) rotated.push_back(alpha[(i + rot) % n]);
    return rotated;
}

std::vector<uint8_t> base47_digits_to_5bytes(const std::vector<int>& digits) {
    uint64_t val = 0;
    for (auto d : digits) val = val * BASE + static_cast<uint64_t>(d);
    std::vector<uint8_t> out(5);
    for (int i = 4; i >= 0; --i) {
        out[i] = val & 0xFF;
        val >>= 8;
    }
    return out;
}

int checksum47(const std::vector<std::vector<uint8_t>>& blocks) {
    int x = 0;
    for (auto const& block : blocks) 
        for (auto b : block) x ^= b;
    return x % BASE;
}

std::vector<uint8_t> decode_mosaic(std::string_view s) {
    auto rev_base = build_rev(ALPHABET);
    std::vector<uint8_t> out;
    std::vector<std::vector<uint8_t>> cs_windows;
    int block_index = 0;
    size_t i = 0;

    while (i < s.size()) {
        while (i < s.size() && std::isspace(static_cast<unsigned char>(s[i]))) i++;
        if (i >= s.size()) break;

        // trailer detection
        if (i + 2 < s.size() && s[i] == TERM && s[i+1] == TERM) {
            char pad_char = s[i+2];
            if (!rev_base.contains(pad_char)) throw std::runtime_error("Invalid trailer pad digit");
            int pad_count = rev_base[pad_char];
            if (pad_count < 0 || static_cast<size_t>(pad_count) > out.size()) throw std::runtime_error("Invalid pad count");
            out.resize(out.size() - pad_count);
            i += 3;
            if (i != s.size()) throw std::runtime_error("Extra data after trailer");
            return out;
        }

        int rot = rotation_for_block(block_index);
        auto rotated = rotate_alpha(ALPHABET, rot);
        auto rev_rot = build_rev(rotated);

        std::vector<int> digits;
        for (size_t k = 0; k < BLOCK_SYMBOLS; ++k) {
            while (i < s.size() && NOISE_SET.contains(s[i])) i++;
            if (i >= s.size()) throw std::runtime_error("Unexpected end of input");
            char c = s[i++];
            if (c == TERM) throw std::runtime_error("Unexpected terminator");
            if (!rev_rot.contains(c)) throw std::runtime_error("Invalid digit character");
            digits.push_back(rev_rot[c]);
        }

        while (i < s.size() && NOISE_SET.contains(s[i])) i++;
        if (i >= s.size() || s[i] != TERM) throw std::runtime_error("Missing block terminator");
        i++;

        auto block5 = base47_digits_to_5bytes(digits);
        out.insert(out.end(), block5.begin(), block5.end());

        cs_windows.push_back(block5);
        if (cs_windows.size() > CHECKSUM_PERIOD)
            cs_windows.erase(cs_windows.begin(), cs_windows.end() - CHECKSUM_PERIOD);

        block_index++;

        if (cs_windows.size() == CHECKSUM_PERIOD) {
            while (i < s.size() && NOISE_SET.contains(s[i])) i++;
            if (i >= s.size()) throw std::runtime_error("Missing checksum character");
            char chk_char = s[i++];
            if (!rev_base.contains(chk_char)) throw std::runtime_error("Invalid checksum char");
            int got = rev_base[chk_char];
            int expect = checksum47(cs_windows);
            if (got != expect) throw std::runtime_error("Checksum mismatch");
            cs_windows.clear();
        }
    }

    throw std::runtime_error("No trailer found; malformed input");
}

std::vector<uint8_t> xor_with_key(const std::vector<uint8_t>& data, std::string_view key) {
    if (key.empty()) return data;
    std::vector<uint8_t> out;
    out.reserve(data.size());
    for (size_t i = 0; i < data.size(); ++i)
        out.push_back(data[i] ^ static_cast<uint8_t>(key[i % key.size()]));
    return out;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <ciphertext> [key]\n";
        return 1;
    }

    std::string_view ciphertext = argv[1];
    std::string_view key = argc >= 3 ? argv[2] : "";

    try {
        auto raw = decode_mosaic(ciphertext);
        raw = xor_with_key(raw, key);

        std::cout << "Decoded bytes (hex): ";
        for (auto b : raw) std::cout << std::hex << std::uppercase << int(b);
        std::cout << "\n";

        std::string text(raw.begin(), raw.end());
        std::cout << "Decoded text (utf-8): " << text << "\n";
    } catch (std::exception& e) {
        std::cerr << "Decoding error: " << e.what() << "\n";
        return 2;
    }
    return 0;
}

File: src/decrypt/swift/decrypt.swift
import Foundation

let ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*_-?"
let NOISE_SET = "abcdefghijklmnopqrstuvwxyz"
let TERM: UInt8 = UInt8(ascii: "~")
let BASE = 47
let BLOCK_BYTES = 5
let BLOCK_SYMBOLS = 8
let CHECKSUM_PERIOD = 4

let REV_BASE: [UInt8: Int] = {
    var rev: [UInt8: Int] = [:]
    for (i, ch) in ALPHABET.utf8.enumerated() { rev[ch] = i }
    return rev
}()

func rotationForBlock(_ blockIndex: Int) -> Int { (blockIndex * 13 + 11) % BASE }

func rotateAlpha(_ alpha: Data, rot: Int) -> Data {
    let n = alpha.count
    let r = ((rot % n) + n) % n
    return alpha[r...] + alpha[..<r]
}

func base47DigitsTo5Bytes(_ digits: [Int]) -> Data {
    var val: UInt64 = 0
    for d in digits { val = val * UInt64(BASE) + UInt64(d) }
    var out = Data(repeating: 0, count: BLOCK_BYTES)
    for i in stride(from: 4, through: 0, by: -1) { out[i] = UInt8(val & 0xFF); val >>= 8 }
    return out
}

func checksum47(_ blocks: [Data]) -> Int {
    blocks.flatMap { $0 }.reduce(0) { $0 ^ Int($1) } % BASE
}

func decodeMosaic(_ s: String) throws -> Data {
    var out = Data()
    var csWindows: [Data] = []
    var blockIndex = 0
    var bytes = Array(s.utf8)
    var i = 0

    while i < bytes.count {
        while i < bytes.count && NOISE_SET.contains(Character(UnicodeScalar(bytes[i]))) { i += 1 }
        if i == bytes.count { break }

        if i + 2 < bytes.count, bytes[i] == TERM, bytes[i+1] == TERM {
            let padChar = bytes[i+2]
            guard let padCount = REV_BASE[padChar], padCount <= out.count else {
                throw NSError(domain: "Invalid trailer or pad mismatch", code: 1)
            }
            if padCount > 0 { out.removeLast(padCount) }
            i += 3
            guard i == bytes.count else { throw NSError(domain: "Extra data after trailer", code: 1) }
            return out
        }

        let rotatedAlpha = rotateAlpha(Data(ALPHABET.utf8), rot: rotationForBlock(blockIndex))
        var revRot: [UInt8: Int] = [:]
        for (idx, ch) in rotatedAlpha.enumerated() { revRot[ch] = idx }

        var digits: [Int] = []
        for _ in 0..<BLOCK_SYMBOLS {
            while i < bytes.count && NOISE_SET.contains(Character(UnicodeScalar(bytes[i]))) { i += 1 }
            guard i < bytes.count, let digit = revRot[bytes[i]] else { throw NSError(domain: "Invalid digit", code: 1) }
            digits.append(digit)
            i += 1
        }

        while i < bytes.count && NOISE_SET.contains(Character(UnicodeScalar(bytes[i]))) { i += 1 }
        guard i < bytes.count, bytes[i] == TERM else { throw NSError(domain: "Missing block terminator", code: 1) }
        i += 1

        let block5 = base47DigitsTo5Bytes(digits)
        out.append(block5)

        csWindows.append(block5)
        if csWindows.count > CHECKSUM_PERIOD { csWindows.removeFirst(csWindows.count - CHECKSUM_PERIOD) }
        blockIndex += 1

        if csWindows.count == CHECKSUM_PERIOD {
            while i < bytes.count && NOISE_SET.contains(Character(UnicodeScalar(bytes[i]))) { i += 1 }
            guard i < bytes.count, let got = REV_BASE[bytes[i]] else { throw NSError(domain: "Invalid checksum", code: 1) }
            i += 1
            let expect = checksum47(csWindows)
            guard got == expect else { throw NSError(domain: "Checksum mismatch", code: 1) }
            csWindows.removeAll()
        }
    }

    throw NSError(domain: "No trailer found; malformed input", code: 1)
}

func main() {
    let args = CommandLine.arguments
    guard args.count >= 2 else {
        print("Usage: \(args[0]) <ciphertext> [key]")
        return
    }

    let ciphertext = args[1]
    let keyData = args.count >= 3 ? Data(args[2].utf8) : Data()

    do {
        var raw = try decodeMosaic(ciphertext)
        if !keyData.isEmpty {
            for i in 0..<raw.count { raw[i] ^= keyData[i % keyData.count] }
        }

        print("Decoded bytes (hex):", raw.map { String(format: "%02X", $0) }.joined())
        if let text = String(data: raw, encoding: .utf8) {
            print("Decoded text (utf-8):", text)
        } else {
            print("Decoded text: (not valid UTF-8)")
        }
    } catch {
        print("Decoding error:", error.localizedDescription)
    }
}

main()
